addons/binding/org.openhab.binding.smappee/src/main/java/org/openhab/binding/smappee/internal/SmappeeService.java

```diff
@@ -0,0 +1,473 @@
+                String nowUtcMinus20MinMillis = String.valueOf(nowUtcMinus20Min.getMillis());
+
+                // sample API method to call :
+                // https://app1pub.smappee.net/dev/v1/servicelocation/123/consumption?aggregation=1&from=1388534400000&to=1391212800000
+
+                String responseReadings = getData("/dev/v1/servicelocation/" + this.serviceLocationId
+                        + "/consumption?aggregation=1&from=" + nowUtcMinus20MinMillis + "&to=" + nowUtcMillis);
+
+                if (responseReadings.isEmpty()) {
+                    return null;
+                }
+
+                SmappeeDeviceReading readings = gson.fromJson(responseReadings, SmappeeDeviceReading.class);
+
+                // Sum of 5 minutes in Wh so average power is 60/5 x
+
+                for (SmappeeDeviceReadingConsumption consumption : readings.consumptions) {
+                    consumption.consumption = consumption.consumption * 12;
+                    consumption.solar = consumption.solar * 12;
+                }
+
+                logger.debug("smappee'{}' read", this.serviceLocationId);
+
+                return readings;
+
+            } catch (CommunicationException se) {
+                logger.error("failed to read smappee '{}'", se.getMessage());
+            } catch (JsonSyntaxException pe) {
+                logger.warn("failed to read response from smappee : {}", pe.getMessage());
+            }
+        } while (currentTry++ < this.retry);
+
+        return null;
+    }
+
+    public SmappeeServiceLocationInfo getServiceLocationInfo() {
+        int currentTry = 0;
+        do {
+            try {
+                // sample API method to call :
+                // https://app1pub.smappee.net/dev/v1/servicelocation/123/info
+
+                String responseReadings = getData("/dev/v1/servicelocation/" + this.serviceLocationId + "/info");
+
+                if (responseReadings.isEmpty()) {
+                    return null;
+                }
+
+                SmappeeServiceLocationInfo readings = gson.fromJson(responseReadings, SmappeeServiceLocationInfo.class);
+
+                logger.debug("servicelocationinfo '{}' read", this.serviceLocationId);
+
+                return readings;
+
+            } catch (CommunicationException se) {
+                logger.warn("failed to read servicelocationinfo '{}'", se.getMessage());
+            } catch (JsonSyntaxException pe) {
+                logger.warn("failed to read response from smappee : {}", pe.getMessage());
+            }
+        } while (currentTry++ < this.retry);
+
+        return null;
+    }
+
+    /**
+     * Gets the smappee energy consumption readings for a specific appliance.
+     *
+     * @return the device readings
+     */
+    public SmappeeApplianceEvent getLatestApplianceReading(String applianceId) {
+        int currentTry = 0;
+        do {
+            try {
+                DateTime nowUtc = DateTime.now(DateTimeZone.UTC).minusMinutes(1);
```

Comment:
I've seen a similar code block above - wouldn't it make sense to extract it to a common, parameterized method?

Reply:
@sjka Do you mean the retry mechanism ?